mmverify.py

```diff
@@ -1,369 +1,588 @@
+    def add_e(self, stmt: Stmt, label: Label) -> None:
+        """Add an essential hypothesis (token tuple) to the frame stack
+        top.
+        """
         frame = self[-1]
-        frame.d.update(((min(x,y), max(x,y))
-                        for x, y in itertools.product(stat, stat) if x != y))
-
-    def lookup_c(self, tok): return any((tok in fr.c for fr in reversed(self)))
-    def lookup_v(self, tok): return any((tok in fr.v for fr in reversed(self)))
-
-    def lookup_f(self, var):
-        for frame in reversed(self):
-            try: return frame.f_labels[var]
-            except KeyError: pass
+        frame.e.append(stmt)
+        frame.e_labels[tuple(stmt)] = label
+        # conversion to tuple since dictionary keys must be hashable
+
+    def add_d(self, varlist: list[Var]) -> None:
+        """Add a disjoint variable condition (ordered pair of variables) to
+        the frame stack top.
+        """
+        self[-1].d.update((min(x, y), max(x, y))
+                          for x, y in itertools.product(varlist, varlist) if x != y)
+
+    def lookup_c(self, tok: Const) -> bool:
+        """Return whether the given token is an active constant."""
+        return any(tok in fr.c for fr in self)
+
+    def lookup_v(self, tok: Var) -> bool:
+        """Return whether the given token is an active variable."""
+        return any(tok in fr.v for fr in self)
+
+    def lookup_d(self, x: Var, y: Var) -> bool:
+        """Return whether the given ordered pair of tokens belongs to an
+        active disjoint variable statement.
+        """
+        return any((min(x, y), max(x, y)) in fr.d for fr in self)
+
+    def lookup_f(self, var: Var) -> Label:
+        """Return the label of the active floating hypothesis which types the
+        given variable.
+        """
+        for frame in self:
+            try:
+                return frame.f_labels[var]
+            except KeyError:
+                pass
         raise MMKeyError(var)
 
-    def lookup_d(self, x, y):
-        return any(((min(x,y), max(x,y)) in fr.d for fr in reversed(self)))
-
-    def lookup_e(self, stmt):
+    def lookup_e(self, stmt: Stmt) -> Label:
+        """Return the label of the (earliest) active essential hypothesis with
+        the given statement.
+        """
         stmt_t = tuple(stmt)
-        for frame in reversed(self):
-            try: return frame.e_labels[stmt_t]
-            except KeyError: pass
+        for frame in self:
+            try:
+                return frame.e_labels[stmt_t]
+            except KeyError:
+                pass
         raise MMKeyError(stmt_t)
 
-    def make_assertion(self, stat):
-        frame = self[-1]
+    def find_vars(self, stmt: Stmt) -> list[Var]:
+        """Return the list of variables in the given statement."""
+        var_list = []
+        for x in stmt:
+            if x not in var_list and self.lookup_v(x):
+                var_list.append(x)
+        return var_list
```

Comment:
Naive question: why not use a set for `var_list`?

Reply:
You're right. One-liner in next commit: ```return {x for x in stmt if self.lookup_v(x)}```